source/hp/fe_collection.cc

```diff
@@ -21,6 +21,68 @@ DEAL_II_NAMESPACE_OPEN
 
 namespace hp
 {
+  /**
+   * Try to find a finite element inside FECollection of @p dof_handler
+   * which dominates FEs given as fe_inices in @p fes .
+   * This function looks for a least dominant element. For example, if
+   * FECollection consists of {Q1,Q2,Q3,Q4} and we are looking for the least
+   * dominant FE for Q3 and Q4 (@p fes is {2,3}), then the least dominant is
+   * Q3 and therefore the return value will be 2.
+   *
+   * If we were not able to find a finite element, the function returns
+   * numbers::invalid_unsigned_int .
+   */
+  template <int dim, int spacedim>
+  unsigned int
+  FECollection<dim,spacedim>::find_least_face_dominating_fe (const std::set<unsigned int> &fes) const
+  {
+    const hp::FECollection<dim,spacedim> &fe_collection = *this;
+    std::set<unsigned int> candidate_fes;
+
+    // first loop over all FEs and check which can dominate those given in FEs:
+    for (unsigned int cur_fe = 0; cur_fe < fe_collection.size(); cur_fe++)
+      {
+        FiniteElementDomination::Domination domination = FiniteElementDomination::no_requirements;
+        // check if cur_fe can dominate all FEs in @p fes:
+        for (std::set<unsigned int>::const_iterator it = fes.begin();
+             it!=fes.end(); ++it)
+          {
+            Assert (*it < fe_collection.size(),
+                    ExcIndexRangeType<unsigned int> (*it, 0, fe_collection.size()));
+            domination = domination &
+                         fe_collection[cur_fe].compare_for_face_domination
+                         (fe_collection[*it]);
+          }
+
+        // if we found dominating element, keep them in a set.
+        if (domination == FiniteElementDomination::this_element_dominates)
+          candidate_fes.insert(cur_fe);
+      }
+
+    // among the ones we found, pick one that is dominated by all others and
+    // thus should represent the largest FE space.
+    if (candidate_fes.size() == 1)
+      {
+        return *candidate_fes.begin();
+      }
+    else
+      for (std::set<unsigned int>::const_iterator it = candidate_fes.begin(); it!=candidate_fes.end(); ++it)
+        {
+          FiniteElementDomination::Domination domination = FiniteElementDomination::no_requirements;
+          for (std::set<unsigned int>::const_iterator ito = candidate_fes.begin(); ito!=candidate_fes.end(); ++ito)
+            {
+              domination = domination &
+                           fe_collection[*ito].compare_for_face_domination(fe_collection[*it]);
+            }
+
+          if (domination == FiniteElementDomination::this_element_dominates)
+            return *it;
```

Comment:
This code stumped me, because you use the same condition `if (... == this_element_dominates)` as above. I read this as "find the most dominant among the candidates". The difference is in your order of arguments in the call in line 75. 

For clarity and readability, since you're here trying to find the *least* dominating element, can you revert the order of elements in line 75 and make this condition here for `...== other_element_dominates`?

Also, what happens if you end up with `domination==no_requirements` or `domination==neither_element_dominates`? The former could happen if you have an FE_Collection that only contains (multiple instances of) FE_Nothings among the candidates, e.g. `{Nothing x Nothing, Nothing x Nothing, Q1 x Nothing, Nothing x Q1}`. The latter could presumably happen if you had `{Q1,Q1,Q2}`. Maybe making additional testcases out of these degenerate situations would make for good checks.